pow/src/test/java/tech/pegasys/artemis/pow/Eth1DataManagerTest.java

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.pow;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.when;
+import static tech.pegasys.artemis.util.config.Constants.ETH1_FOLLOW_DISTANCE;
+import static tech.pegasys.artemis.util.config.Constants.ETH1_REQUEST_BUFFER;
+import static tech.pegasys.artemis.util.config.Constants.SECONDS_PER_ETH1_BLOCK;
+import static tech.pegasys.artemis.util.config.Constants.SECONDS_PER_SLOT;
+import static tech.pegasys.artemis.util.config.Constants.SLOTS_PER_ETH1_VOTING_PERIOD;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.common.primitives.UnsignedLong;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.web3j.protocol.Web3j;
+import org.web3j.protocol.core.DefaultBlockParameter;
+import org.web3j.protocol.core.DefaultBlockParameterName;
+import org.web3j.protocol.core.Request;
+import org.web3j.protocol.core.methods.response.EthBlock;
+import tech.pegasys.artemis.pow.event.CacheEth1BlockEvent;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.AsyncRunnerTest;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.config.Constants;
+import tech.pegasys.artemis.util.time.TimeProvider;
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class Eth1DataManagerTest {
+
+  private final TimeProvider timeProvider = mock(TimeProvider.class);
```

Comment:
I'd use a stub here for simplicity rather than a mock.  If this was the only place we were likely to need it, a mock would be fine but we're going to mock `TimeProvider` a lot so it makes sense to introduce a stub for it in test-support.  Something like:

```
public class StubTimeProvider extends TimeProvider {
  private UnsignedLong timeInSeconds;

  public StubTimeProvider() {
    this(UnsignedLong.valueOf(29842948));
  }

  public StubTimeProvider(final long timeInSeconds) {
    this(UnsignedLong.valueOf(timeInSeconds));
  }

  public StubTimeProvider(final UnsignedLong timeInSeconds) {
    this.timeInSeconds = timeInSeconds;
  }

  public void advanceTimeBySeconds(final long seconds) {
    this.timeInSeconds = timeInSeconds.plus(UnsignedLong.valueOf(seconds));
  }

  @Override
  public UnsignedLong getTimeInSeconds() {
    return timeInSeconds;
  }
}
```

Reply:
Done.